scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+	return taskMetadata, nil
+}
+
+func (m *TaskMetadataManager) GetAllTasks() (TaskMetadataList, fferr.GRPCError) {
+	metadata, err := m.storage.List(TaskMetadataKey{}.String())
+	if err != nil {
+		return TaskMetadataList{}, err
+	}
+
+	tml, err := m.getAllTasksAsMetadataList(metadata)
+	if err != nil {
+		return TaskMetadataList{}, err
+	}
+	return tml, nil
+}
+
+func (m *TaskMetadataManager) getAllTasksAsMetadataList(metadata map[string]string) (TaskMetadataList, fferr.GRPCError) {
+	tml := TaskMetadataList{}
+	for _, meta := range metadata {
+		taskMetadata := TaskMetadata{}
+		err := taskMetadata.Unmarshal([]byte(meta))
+		if err != nil {
+			return TaskMetadataList{}, err
+		}
+		tml = append(tml, taskMetadata)
+	}
+	return tml, nil
+}
+
+func (m *TaskMetadataManager) CreateTaskRun(name string, taskID TaskID, trigger Trigger) (TaskRunMetadata, fferr.GRPCError) {
+	// ids will be generated by TM
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task name", name)
+		return TaskRunMetadata{}, err
+	}
+
+	// Not sold on this naming for this struct. Maybe like RunHistory or something?
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	latestID, err := runs.GetLatestRunId()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	startTime := time.Now().UTC()
+
+	metadata := TaskRunMetadata{
+		ID:          TaskRunID(latestID + 1),
+		TaskId:      taskID,
+		Name:        name,
+		Trigger:     trigger,
+		TriggerType: trigger.Type(),
+		Status:      Pending,
+		StartTime:   startTime,
+	}
+
+	runs.Runs = append(runs.Runs, TaskRunSimple{RunID: metadata.ID, DateCreated: startTime})
+
+	serializedRuns, err := runs.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+	err = m.storage.Create(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRunMetaKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: startTime}
+	err = m.storage.Create(taskRunMetaKey.String(), string(serializedMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
```

Comment:
i feel like this needs to be transactional. What happens if this part fails, the above was stored but the metadat failed